### DISCUSSION TITLE
How can tox configuration set virtualenv discovery?

### DISCUSSION THREAD

**Discussion by bluss** [tox-dev/tox#3029] [Q&A] [answered]
Virtualenv discovery is the method which tox 4 uses to find python interpreters.

[Rye](https://rye-up.com) is a tool in development (probably not ready for production) which pulls and installs multiple python verions, so obviously you'd want to get it integrated with tox ([gh issue @ rye](https://github.com/mitsuhiko/rye/issues/206)).

I've made a try to get tox to get python versions from Rye and "it works" ([code here](https://github.com/bluss/virtualenv-rye-discovery)). I need to set `VIRTUALENV_DISCOVERY=rye` for this to be active. Is there a good way to configure this in tox? It's an environment variable that needs to be active before it creates specific test environments.

A. You can set the variable explicitly on the command line  
B. You can "monkey patch" the environment variable in toxfile.py, that also works. Is it good enough? I would like to have a more official way:

```python
import logging
import os

from tox.tox_env.register import ToxEnvRegister
from tox.plugin import impl

@impl
def tox_register_tox_env(register: ToxEnvRegister):
    logging.getLogger(__name__).warning("Use virtualenv discovery = Rye")
    os.environ['VIRTUALENV_DISCOVERY'] = 'rye'
```

**Comment by gaborbernat** [comment, marked as answer]
You should be setting the env var via a plugin, as show in https://github.com/tox-dev/tox/blob/b138c5cf87f49a216fea9e7f507f96da1b9fe2b7/tests/plugin/test_plugin.py#L212-L229

**Reply by bluss** [reply]
Thanks a lot, that seems to do the right thing, and it works well :slightly_smiling_face:

**Reply by bluss** [reply]
The WIP plugin that comes out of this lives @ https://github.com/bluss/tox-rye for the moment
